### ISSUE TITLE
Fix implementation of `RandomNumberExtension` in the User Guide

### ISSUE THREAD

**Issue by graydavid** [junit-team/junit5#3433]
## Steps to reproduce

Copying the relevant parts of the `RandomNumberExtension` example in the documentation [here](https://junit.org/junit5/docs/snapshot/user-guide/index.html#extensions-RandomNumberExtension):

```java
class RandomNumberDemo {

    // Use static randomNumber0 field anywhere in the test class,
    // including @BeforeAll or @AfterEach lifecycle methods.
    @Random
    private static Integer randomNumber0;

    // Use randomNumber1 field in test methods and @BeforeEach
    // or @AfterEach lifecycle methods.
    @Random
    private int randomNumber1;
...
}

class RandomNumberExtension
        implements BeforeAllCallback, TestInstancePostProcessor, ParameterResolver {

...
    private void injectFields(Class<?> testClass, Object testInstance,
            Predicate<Field> predicate) {

        predicate = predicate.and(field -> isInteger(field.getType()));
        findAnnotatedFields(testClass, Random.class, predicate)
            .forEach(field -> {
                try {
                    field.setAccessible(true);
                    field.set(testInstance, this.random.nextInt());
                }
                catch (Exception ex) {
                    throw new RuntimeException(ex);
                }
            });
    }
...
}
```

There are a couple of issues with this code:

1. `randomNumber0` is an Integer, so it's not injected.
2. `findAnnotatedFields` is missing the fourth parameter, so it doesn't compile.
3. If `randomNumber0` is removed completely from `RandomNumberDemo`, then `randomNumber1` is not injected. I believe the root cause is that `TestMethodTestDescriptor` does not invoke `TestInstancePostProcessor` (side note: even though it does invoke `TestInstancePreDestroyCallback`). Meanwhile, `ClassBasedTestDescriptor` does, which is why `randomNumber1` is populated when `randomNumber0` is present. I can't tell which part is wrong here: the documentation or the code.

Note: The code in https://github.com/junit-team/junit5/issues/3004#issuecomment-1215045050 fixes all these issues. The last issue is resolved by using `BeforeEachCallback` instead of `TestInstancePostProcessor`.

## Context

 - Used versions (Jupiter/Vintage/Platform): 5.9.3
 - Build Tool/IDE: Eclipse 2023-03 (4.27.0)

## Deliverables

- <strike>Remove bugs from documentation</strike>
- <strike>Possibly change `TestMethodTestDescriptor` to start invoking `TestInstancePostProcessor` (if that's desirable)... or switch the example to use `BeforeEachCallback` (and mention the limitations with `TestMethodTestDescriptor` and `TestInstancePostProcessor`)</strike>
- [x] Fix implementation of `isInteger()` in `RandomNumberExtension` in the User Guide.

**Comment by sbrannen** [comment]
Hi @graydavid,

Thanks for reporting the issues you discovered, and congratulations on submitting your first issue for JUnit 5! 👍 

I will fix the issues with the implementation of the `RandomNumberExtension`

**Comment by sbrannen** [comment]
> 1. `randomNumber0` is an Integer, so it's not injected.

Good catch. I've fixed that locally and will push that soon.

> 2. `findAnnotatedFields` is missing the fourth parameter, so it doesn't compile.

The example is in the documentation for JUnit 5.10, and `org.junit.platform.commons.support.AnnotationSupport.findAnnotatedFields(Class<?>, Class<? extends Annotation>, Predicate<Field>)` was introduced in JUnit Platform 1.10. So, the code compiles fine.

Please note that almost all demos in the User Guide are actually compiled as part of the but JUnit 5 build. See the `documentation` project for details.

> 3. If `randomNumber0` is removed completely from `RandomNumberDemo`, then `randomNumber1` is not injected. 

I cannot reproduce that. Based my tests, `randomNumber1` is always injected.

**Comment by graydavid** [comment]
Thanks for the quick response!

> was introduced in JUnit Platform 1.10. So, the code compiles fine.

My apologies. I thought I had updated to the latest version, but I was one behind.

> Please note that almost all demos in the User Guide are actually compiled as part of the but JUnit 5 build. See the documentation project for details.

That's awesome!

> I cannot reproduce that. Based my tests, randomNumber1 is always injected.

I didn't give complete instructions. Doing more testing, you also have to remove the `@Random` parameter to test method as well.

Here's a repo with a minimal reproduction case: https://github.com/graydavid/junit-random-demo

The most important part is `RandomNumberDemoTest `, which fails unless there's either an `@Random` static field or method parameter:

```
public class RandomNumberDemoTest {

  @Random
  private int randomNumber1;

  @Test
  void test() {
    assertNotEquals(0, randomNumber1);
  }
}
```

Here's more information on my runtime environment:

```
Apache Maven 3.8.7 (b89d5959fcde851dcb1c8946a785a163f14e1e29)
Java version: 1.8.0_333, vendor: Oracle Corporation, runtime: ...
```

I did also try against java 17 as well.

It doesn't seem like I can reopen this issue, so I'll tag you and hope that gets a response: @sbrannen. If that doesn't work by Tuesday, I'll open a new issue and point to this comment.

**Comment by graydavid** [comment]
Also, just a note: because of the issue I was seeing with `TestInstancePostProcessor`, even though the junit documentation says it's appropriate for injecting dependencies, I switched to the approach that mockito is using (see [MockitoExtension](https://github.com/mockito/mockito/blob/eeeff29fabddd56c26d8e0436e9c523419a1e3d0/subprojects/junit-jupiter/src/main/java/org/mockito/junit/jupiter/MockitoExtension.java#L118)), which relies on `BeforeEachCallback` and `AfterEachCallback` instead.

**Comment by sbrannen** [comment]
> It doesn't seem like I can reopen this issue, so I'll tag you and hope that gets a response: @sbrannen. If that doesn't work by Tuesday, I'll open a new issue and point to this comment.

Thanks for all of the feedback. Much appreciated! 👍 

I've opened #3437 to address the underlying issue in JUnit Jupiter.

**Comment by sbrannen** [comment]
Reopening to ensure the example works against `5.10.x` since #3437 will not be addressed until 5.11.
